expression/builtin_other.go

```diff
@@ -215,3 +217,30 @@ func (b *builtinBitCountSig) eval(row []types.Datum) (d types.Datum, err error)
 	d.SetInt64(count)
 	return d, nil
 }
+
+// for plan cache of prepared statements
+type getParamFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *getParamFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	bt := &builtinGetParamSig{newBaseBuiltinFunc(args, ctx)}
+	bt.deterministic = false
+	bt.deferred = true
+	return bt.setSelf(bt), errors.Trace(c.verifyArgs(args))
+}
+
+type builtinGetParamSig struct {
+	baseBuiltinFunc
+}
+
+func (b *builtinGetParamSig) eval(row []types.Datum) (types.Datum, error) {
```

Comment:
since we a rewriting expression, the old API `eval()` should be abandoned after refactoring, you should rewrite this function using the new expression evaluation framework, see #4080 for examples.

Reply:
Thank you for the sugguestion. I've changed the codes as follows. As I mentioned above, more typed functions will be added when typed parameters are supported.

`-type builtinGetParamSig struct {`
`-  baseBuiltinFunc`
`+type builtinGetParamStringSig struct {`
`+  baseStringBuiltinFunc`
` }`
` `
`-func (b *builtinGetParamSig) eval(row []types.Datum) (types.Datum, error) {`
`+func (b *builtinGetParamStringSig) evalString(row []types.Datum) (string, bool, error) {`
`   args, err := b.evalArgs(row)`
`   if err != nil {`
`-      return types.Datum{}, errors.Trace(err)`
`+      return "", true, errors.Trace(err)`
`   }`
`   idx := args[0].GetInt64()`
`   sessionVars := b.ctx.GetSessionVars()`
`-  v := sessionVars.Params[idx]`
`-  return v.(types.Datum), nil`
`+  v := sessionVars.PreparedParams[idx]`
`+`
`+  dt := v.(types.Datum)`
`+  str, err := (&dt).ToString()`
`+  if err != nil {`
`+      return "", true, errors.Trace(err)`
`+  }`
`+  return str, false, nil`
` }`